### DISCUSSION TITLE
Handling paginated APIs with Mutiny

### DISCUSSION THREAD

**Discussion by giscus** [quarkusio/quarkus#25107] [Quarkus Blog/Website]
# Handling paginated APIs with Mutiny

Quarkus: Supersonic Subatomic Java

https://quarkus.io/blog/mutiny-pagination/

**Comment by cobar79** [comment]
Took a couple of days to find this, but exactly what I needed for https://ucdpapi.pcr.uu.se/api/gedevents/22.1?pagesize=1000

**Comment by cobar79** [comment]
How would you convert this to concurrently pull a configurable amount of pages?
In my case https://ucdpapi.pcr.uu.se/api/gedevents/22.1?pagesize=1000 is 262 pages and 261K worth of details.

**Comment by tomschulze** [comment]
Thanks for the write-up! This helped me a lot for retrieving results from a GraphQL API that uses cursor-based pagination.  [Tmarven's stackoverflow answer](https://stackoverflow.com/questions/69815030/is-recursion-necessary-in-quarkus-mutiny) helped me retrieving the cursor from the response.

**Comment by IvanPuntev** [comment]
Here is an example how to handle different pagination https://github.com/quarkusio/quarkus/discussions/36681

**Comment by Marcuse7** [comment]
The hosted Punk API will shut down May 1<sup>st</sup>, 2024. You could run it [locally](https://github.com/sammdec/punkapi-server?tab=readme-ov-file), though.

**Comment by Darkvater** [comment]
Well, the example is handy, but if you actually dig deeper you see it has a performance penalty. To retrieve all pages this implementation always does **1** extra call as it only knows when to stop when the size of beers are finished.

This could be fine in some scenarios, but in others where you know the page size you can optimise it as you can already stop when returned `beers < page size` as you know that is the last page. To do this one should change the stop condition from `.until(List::isEmpty)` to `.whilst( it -> it.size >= pageSize)`. However in the above scenario for some reason Mutiny has already scheduled the call to the second page even **before** the first page was retrieved (and then dropping it). The outcome is the same, but we did do one extra call still; at least on smallrye.reactive.mutiny 2.9.4.

Would be great to see an example which works properly knowing the page size in advance and doign the optimal number of calls.
